### PR TITLE
Created services to allow creation of new address

### PR DESCRIPTION
**Pull request by hamza-yusuff**
## Notion ticket link

[Create Service to add new address](https://www.notion.so/uwblueprintexecs/Create-service-to-add-new-address-350083b689fd47e6998d4b65d82d4786)

## Implementation description
* Created class, and the associated method to allow creation of a new address record in the db when a CreateAddressDTO is passed to the method. AddressDTO is returned if the function is successful otherwise exception is raised.
* Added corresponding DTO classes with validation methods.
* Added tests to verify functionality

## Steps to test
1) Upgrade your database docker exec -it cas_py_backend /bin/bash -c "flask db upgrade"
2) Run tests: docker exec -it cas_py_backend /bin/bash -c "pip install -e . && pytest"
3) Verify that all of the tests pass

## What should reviewers focus on?
* Do all the tests pass

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/python/tests/functional/test_address_service.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        # boundary case with city being assigned an empty string
+        CreateAddressDTO(
+            street_address="333 Lester",
+            city="",
+            postal_code="N2L 3W6",
+            latitude=8.380900,
+            longitude=9.899900,
+        ),
+    ]
+    test_address_service = AddressService(current_app.logger)
```
Comment (jennchenn) [inline backend/python/tests/functional/test_address_service.py:43]: similar to how it's done in the other files right now, can you move this out into a fixture at the top of the file?
https://github.com/uwblueprint/childrens-aid-society/blob/0258555311c572b1d58c18becba0bb5d4a8d4667/backend/python/tests/functional/test_user_service.py#L24-L28

File: backend/python/tests/functional/test_address_service.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            street_address="333 Lester",
+            city="Waterloo",
+            postal_code="N2L 3W6",
+            latitude=8.380900,
+            longitude=9.899900,
+        ),
+        # boundary case with city being assigned an empty string
+        CreateAddressDTO(
+            street_address="333 Lester",
+            city="",
```
Comment (jennchenn) [inline backend/python/tests/functional/test_address_service.py:37]: this is an interesting case - we actually _don't_ want empty strings for `street_address`, `city`, `postal_code`
your validation currently doesn't handle this, but at the very least i don't think we should include this under the valid cases

File: backend/python/tests/functional/test_address_service.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import pytest
+from flask import current_app
+
+from app.resources.address_dto import AddressDTO, CreateAddressDTO
+from app.services.implementations.address_service import AddressService
+
+
+def test_create_address_success():
```
Comment (jennchenn) [inline backend/python/tests/functional/test_address_service.py:8]: can you actually separate each of these into their own tests rather than running them in one batch (this applies for both the valid and invalid test cases)?

in the case of errors, it'll make it easier to pinpoint which test case is causing the failure if we have them isolated in their own tests (additionally this separates out the testing so one failed test case doesn't cause all of them to not be run)

as a bonus, if we implement test code coverage reporting, having the test cases separated out will be useful too!

File: backend/python/tests/functional/test_address_service.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    valid_cases = [
+        # normal case
```
Comment (jennchenn) [inline backend/python/tests/functional/test_address_service.py:10]: for the normal case, you could also validate the fields in the return object match the input (not necessary for all the test cases, but maybe just one)

File: backend/python/tests/functional/test_address_service.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        ),
+        # boundary case with city being assigned an empty string
+        CreateAddressDTO(
+            street_address="333 Lester",
+            city="",
+            postal_code="N2L 3W6",
+            latitude=8.380900,
+            longitude=9.899900,
+        ),
+    ]
```
Comment (jennchenn) [inline backend/python/tests/functional/test_address_service.py:42]: since `latitude` and `longitude` are the only non-nullable fields right now, can you add a test case where neither are provided to ensure it works?

File: backend/python/tests/functional/test_address_service.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    test_address_service = AddressService(current_app.logger)
+    for valid in valid_cases:
+        address_instance = test_address_service.create_address(valid)
+        assert type(address_instance) is AddressDTO
+
+
+def test_create_address_invalid_field_fails():
```
Comment (jennchenn) [inline backend/python/tests/functional/test_address_service.py:49]: since you added validation for an empty address being passed to the service, you can add that test case here!

```
with pytest.raises(Exception):
     test_address_service.create_address(None)
```

**Review by jennchenn — CHANGES_REQUESTED** [review]
nice work thinking up some unique test cases!!

added some thoughts + suggestions for potential refactors/testing improvements

File: backend/python/app/resources/address_dto.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+class AddressDTO:
+    def __init__(self, id, street_address, city, postal_code, latitude, longitude):
+        self.id = id
```
[the remark this answers is not in the document]
Comment (hamza-yusuff) [reply backend/python/app/resources/address_dto.py:3]: Done!

File: backend/python/app/resources/address_dto.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        self.longitude = longitude
+
+
+class CreateAddressDTO:
+    def __init__(self, **kwargs):
+        self.street_address = kwargs.get("street_address")
+        self.city = kwargs.get("city")
+        self.postal_code = kwargs.get("postal_code")
+        self.latitude = kwargs.get("latitude")
+        self.longitude = kwargs.get("longitude")
```
[the remark this answers is not in the document]
Comment (hamza-yusuff) [reply backend/python/app/resources/address_dto.py:17]: Made the changes!

File: backend/python/tests/functional/test_address_service.py (hunk 3 above)
Comment (hamza-yusuff) [reply backend/python/tests/functional/test_address_service.py:8]: Just a bit confused on this. Do i separate the cases into their own function? for example functions like - test_create_address_success_normal_case(), test_create_address_success_boundary_case(), test_create_address_success_precision_case() - for each test case?

File: backend/python/app/services/implementations/address_service.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+from ..interfaces.address_service import IAddressService
+
+
+class AddressService(IAddressService):
+    def __init__(self, logger):
+        self.logger = logger
+
+    def create_address(self, address):
+        try:
+            if address:
```
[the remark this answers is not in the document]
Comment (hamza-yusuff) [reply backend/python/app/services/implementations/address_service.py:13]: Yess, changed the code to log the error only once, also added checks to see if it is of type CreateDTO or not

File: backend/python/tests/functional/test_address_service.py (hunk 3 above)
Comment (jennchenn) [reply backend/python/tests/functional/test_address_service.py:8]: Yes - you can test more than once for normal case, but I think this makes it clearer when we analyze the test results after the tests are run (i.e. instead of seeing `test_create_address_success` fail and having to check which of the specific cases caused it to fail, we can pinpoint that it's a boundary case or precision, etc. that caused the failure)

^ you don't have to separate it out if it seems pretty similar to you, but i think if you have different conditions/things you're testing, would be good to separate it out

File: backend/python/tests/functional/test_address_service.py (hunk 1 above)
Comment (hamza-yusuff) [reply backend/python/tests/functional/test_address_service.py:43]: Done!

File: backend/python/tests/functional/test_address_service.py (hunk 3 above)
Comment (hamza-yusuff) [reply backend/python/tests/functional/test_address_service.py:8]: I created tow separate class for the normal and invalid cases, and then put the test cases as methods for the classes to keep the cases grouped according to what they test essentially.

File: backend/python/tests/functional/test_address_service.py (hunk 4 above)
Comment (hamza-yusuff) [reply backend/python/tests/functional/test_address_service.py:10]: Yess I have put a test case which validates the return object's field with the input's

File: backend/python/tests/functional/test_address_service.py (hunk 5 above)
Comment (hamza-yusuff) [reply backend/python/tests/functional/test_address_service.py:42]: Yeah added this well

**Review by jennchenn — APPROVED** [review]
i think this looks good!! nice 🥳
